### PR TITLE
[WEB-738] implement api tests

### PR DESCRIPTION
**Pull request by ginnyyadav**
Hey Jamie, 

this is almost done. I want to add more tests from the custodial user perspective, but that aspect is currently blocked while backend investigates the issues with the confirm/send/signup endpoint.

That being said there is ALOT added here, so I wanted to create a pull request because this is going to take some time to review. I'm not sure if you'd like to review this together or you'd rather review it on your own and sync up later. let me know.

some aspects I'd like you to think about 

- i've created a variable for uploadId the mimics what is created by the uploader. I recently had a discussion with lenny on how uploads shouldn't have the same upload ID, which is why I turned this into a variable.

- The delete user request is not added in these tests because it is only semi-functional at the moment

- Added bolus and cgb data to the uploads for a user and their corresponsding schemas. Thinking about also adding basal data as well since its another "main" data type.

- added some tests to "clean up" and "reset" the environments afterwards (removing the data donation selection, removing the clinican from sharing)

### CODE REVIEW

**Comment by Benderr-TP** [comment]
I didn’t know you were using Postman @ginnyyadav , will have to sync up collections 😁

File: tests/API_smoke_test.postman_collection.json [hunk 1]
```diff
@@ -464,6 +465,3139 @@
...
+							"name": "Clinician sign up",
+							"event": [
+								{
+									"listen": "prerequest",
+									"script": {
+										"id": "72e6dbe1-2283-44b9-8e51-76fbab62e71a",
+										"exec": [
+											"var moment = require('moment');",
+											"",
+											"    pm.variables.set(\"currentTimestamp\", moment().unix());",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: [nit] blank space

File: tests/API_smoke_test.postman_collection.json [hunk 2]
```diff
@@ -464,6 +465,3139 @@
...
+						{
+							"name": "Clinician sign up",
+							"event": [
+								{
+									"listen": "prerequest",
+									"script": {
+										"id": "72e6dbe1-2283-44b9-8e51-76fbab62e71a",
+										"exec": [
+											"var moment = require('moment');",
+											"",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: [nit] extra lines

File: tests/API_smoke_test.postman_collection.json [hunk 3]
```diff
@@ -464,6 +465,3139 @@
...
+									"listen": "test",
+									"script": {
+										"id": "aef68e59-fa85-4675-bd0c-d73bcfc4f666",
+										"exec": [
+											"const assert = eval(globals.loadAsserts);",
+											"assert.common(201, 1000, true, null,\"\");",
+											"let jsonData = pm.response.json();",
+											"pm.variables.set(\"sessionToken\", pm.response.headers.get(\"x-tidepool-session-token\"));",
+											"pm.variables.set(\"userId\", jsonData.userid);",
+											"",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: [nit] white space

File: tests/API_smoke_test.postman_collection.json [hunk 4]
```diff
@@ -464,6 +465,3139 @@
...
+							"name": "Send email verification",
+							"event": [
+								{
+									"listen": "test",
+									"script": {
+										"id": "0fb5a1e8-5b9d-45d1-ac8c-0c0f0352ee5e",
+										"exec": [
+											"const assert = eval(globals.loadAsserts);",
+											"assert.common(200, 1000, false, null,\"\");",
+											"    ",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: [nit] white space

File: tests/API_smoke_test.postman_collection.json [hunk 5]
```diff
@@ -464,6 +465,3139 @@
...
+											"let now = moment();",
+											"pm.variables.set(\"uploadDeviceDataItemTime\", now.format());",
+											"pm.variables.set(\"uploadDeviceDataItemDeviceTime\", now.format('YYYY-MM-DDThh:mm:ss'));",
+											"",
+											"function randomString(length, chars) {",
+											"    var result = '';",
+											"    for (var i = length; i > 0; --i) result += chars[Math.round(Math.random() * (chars.length - 1))];",
+											"    return result;",
+											"}",
+											"",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: I think you can shorten this by using similar / the same to here https://github.com/tidepool-org/APITests/blob/d22349a8a40ce3cce194f47edcdd0a45310b1dca/tests/API_smoke_test.postman_collection.json#L271

File: tests/API_smoke_test.postman_collection.json [hunk 6]
```diff
@@ -464,6 +465,3139 @@
...
+									"listen": "test",
+									"script": {
+										"id": "aef68e59-fa85-4675-bd0c-d73bcfc4f666",
+										"exec": [
+											"const assert = eval(globals.loadAsserts);",
+											"assert.common(201, 1000, true, null,\"\");",
+											"let jsonData = pm.response.json();",
+											"pm.variables.set(\"sessionToken\", pm.response.headers.get(\"x-tidepool-session-token\"));",
+											"pm.variables.set(\"userId\", jsonData.userid);",
+											"",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: [nit] white space

File: tests/API_smoke_test.postman_collection.json [hunk 7]
```diff
@@ -464,6 +465,3139 @@
...
+								},
+								{
+									"listen": "prerequest",
+									"script": {
+										"id": "41dbe3a9-efca-4b72-a543-b58ac98dc070",
+										"exec": [
+											"var moment = require('moment');\r",
+											"\r",
+											"    pm.variables.set(\"currentTimestamp\", moment().format());"
+										],
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: [nit] formatting / white space

File: tests/API_smoke_test.postman_collection.json [hunk 8]
```diff
@@ -464,6 +465,3139 @@
...
+								"header": [
+									{
+										"key": "x-tidepool-session-token",
+										"value": "{{sessionToken}}",
+										"type": "text"
+									}
+								],
+								"body": {
+									"mode": "raw",
+									"raw": "{\"email\":\"bigdata@tidepool.org\",\"permissions\":{\"view\":{},\"note\":{}}}",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: q: email as a variable or does it not matter?? I have no opinion either way just wanted to note it

File: tests/API_smoke_test.postman_collection.json [hunk 9]
```diff
@@ -464,6 +465,3139 @@
...
+								"method": "PUT",
+								"header": [
+									{
+										"key": "x-tidepool-session-token",
+										"value": "{{sessionToken}}",
+										"type": "text"
+									}
+								],
+								"url": {
+									"raw": "{{api}}/confirm/{{userId}}/invited/bigdata@tidepool.org",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: hmmm - that email again. maybe a variable? :)

File: tests/API_smoke_test.postman_collection.json [hunk 10]
```diff
@@ -464,6 +465,3139 @@
...
+											"const moment = require('moment');",
+											"let now = moment();",
+											"pm.variables.set(\"uploadDeviceDataItemTime\", now.format());",
+											"pm.variables.set(\"uploadDeviceDataItemDeviceTime\", now.format('YYYY-MM-DDThh:mm:ss'));",
+											"",
+											"function randomString(length, chars) {",
+											"    var result = '';",
+											"    for (var i = length; i > 0; --i) result += chars[Math.round(Math.random() * (chars.length - 1))];",
+											"    return result;",
+											"}",
```
Comment (jh-bate) [inline tests/API_smoke_test.postman_collection.json]: random string function again that should be able to use https://github.com/tidepool-org/APITests/blob/d22349a8a40ce3cce194f47edcdd0a45310b1dca/tests/API_smoke_test.postman_collection.json#L271

**Review by jh-bate — APPROVED** [review]
there are two(?) instances of a random string function that I think can be simplified to use an exiting function and a few small formatting cleanup items otherwise LGTM

**Comment by ginnyyadav** [comment]
> I didn’t know you were using Postman @ginnyyadav , will have to sync up collections 😁

Jamie introduced me and I'm in love with it

File: tests/API_smoke_test.postman_collection.json (hunk 5 above)
Comment (ginnyyadav) [reply tests/API_smoke_test.postman_collection.json]: good suggestion. replaced with `pm.variables.set("uploadId", uuid.v4().replace(/-/g, '').slice(0,12));`  in order to simulate a closer version of the upload id we create

**Pull request merged**